### ISSUE TITLE
Finetuning doesn't produce better results

### ISSUE THREAD

**Issue by AlexSteveChungAlvarez** [DigitalPhonetics/IMS-Toucan#189]
Tried finetuning on MLS_portuguese and resulting model can't even synthesize a word. Only produces noise! @Flux9665 I sent you privately the output.

**Comment by Flux9665** [comment]
Did you change the language ID during training an inference appropriately? It will forget all other languages during finetuning and produce noise instead when given a different language embedding.

**Comment by AlexSteveChungAlvarez** [comment]
I did, I sent to your gmail the results and the recipe, here is the piece of the code of the recipe:
```
 # ========================
    # =    Portuguese Data   =
    # ========================
    portuguese_datasets = list()
    chunk_count = 10
    chunks = split_dictionary_into_chunks(build_path_to_transcript_mls_portuguese(), split_n=chunk_count)
    for index in range(chunk_count):
        if index > 4:
            break
        portuguese_datasets.append(prepare_tts_corpus(transcript_dict=chunks[index],
                                                      corpus_dir=os.path.join(PREPROCESSING_DIR, f"mls_portuguese_chunk_{index}"),
                                                      lang="por"))

    all_train_sets.append(ConcatDataset(portuguese_datasets))

    model = ToucanTTS()
```

And here is for the file generation:
```
def portuguese_test(version, model_id, exec_device, speaker_reference=None):
    os.makedirs("audios", exist_ok=True)
    
    read_texts(model_id=model_id,
               sentence=["manda ai que vejo, Vou dar uma saidinha rápida, volto em uma hora, Ai te do atenção!"],
               filename=f"audios/{model_id}_portuguese_test_{version}.wav",
               device=exec_device,
               language="por",
               speaker_reference=speaker_reference
               )
```

```
exec_device = "cuda" if torch.cuda.is_available() else "cpu"
    print(f"running on {exec_device}")

    merged_speaker_references = ["audios/speaker_references/" + ref for ref in os.listdir("audios/speaker_references/")]
    portuguese_test(version="v1",
                    model_id="Meta",
                    exec_device=exec_device,
                    speaker_reference=merged_speaker_references if merged_speaker_references != [] else None)
    
    portuguese_test(version="v1",
                    model_id="Finetuning_Portuguese",
                    exec_device=exec_device,
                    speaker_reference=merged_speaker_references if merged_speaker_references != [] else None)
```

At the beginning I thought it was because I was using the multilingual recipe for 1 language, but then I added LibrittsR_train_clean_100 and still didn't work. What else could it be? @Flux9665

**Comment by Flux9665** [comment]
Then how do the losses look like during finetuning and how do the training progress visualizations look like? Maybe there are faulty datapoints in the finetuning data that could be removed with the scorer?

**Comment by Flux9665** [comment]
Your finetuning script looks fine, except that you are creating the progress plots for french, but that doesn't affect the training in any way.

**Comment by AlexSteveChungAlvarez** [comment]
> Then how do the losses look like during finetuning and how do the training progress visualizations look like? Maybe there are faulty datapoints in the finetuning data that could be removed with the scorer?
![image](https://github.com/user-attachments/assets/1c4221fe-0380-4e11-b44f-2db4ca890bf8)



About faulty datapoints...the dataset is the MLS_portuguese, which you also had included in your stage1 recipe for training from scratch, so if you had to run the scorer for this dataset before training on it, then it would be the case. You have already used it for the main model.
About the progress plots for another language, at the beginning I wanted to include more languages, but as first try on training I only did with a small set to see how it's working now.

**Comment by Flux9665** [comment]
Yes I cleaned all datasets I used with the scorer first, there were plenty of bad samples. 

The progress plots are unfortunatley meaningless in your case, because you generated them for French, but the French language embedding is not being trained, so the TTS will get worse on French as you train.

Are you using wandb? The logging of the individual losses there can be very helpful for finding issues. Generally I'd recommend using the scorer to clean the data first, if you print the top 500 worst samples and their loss, you can see what a good threshold would be for excluding bad samples.

**Comment by AlexSteveChungAlvarez** [comment]
@Flux9665 is it possible to use the scorer before having the model and the cache files? Because your run_scorer.py instructions say they are required, but you are saying that you cleaned all the datasets first. If not, then every time anyone would need to run training twices, first to get these files to run the scorer, and second to actually train. Finetuning on MLS_portuguese worked after running the scorer.

About the French language disappearing, this didn't happen in the previous version of Toucan. Remember when I trained on quechua the first time? It only got the accent but the languages remained on the model, besides, when I trained on quechua + another language, all the languages remained with their accents. Why does this language disappearing when not included in training happen now, as when training yourtts?

**Comment by AlexSteveChungAlvarez** [comment]
It is not possible to run the scorer before having the model and the cache files by now. Already tested it. So everyone needs to train twice if they face this issue with their data.
About language disappearing, I hope this can be fixed in a later release, because it would be amazing to only add new languages without losing previous ones (this was possible with the previous release of toucan, v2)

**Comment by Flux9665** [comment]
It's probably because we use a different decoder now that offers mich better quality, but is a bit more sensitive to changes in the input. In order to keep the performance on an old language, you need to mix in some data from the old language. 

You can run the scorer using the pretrained meta-checkpoint before training your own model.

**Comment by AlexSteveChungAlvarez** [comment]
`You can run the scorer using the pretrained meta-checkpoint before training your own model.`
Would this work even with new datasets not seen in the meta-checkpoint?

**Comment by Flux9665** [comment]
Yes

**Comment by AlexSteveChungAlvarez** [comment]
Now I tested on another dataset of 3 languages (spa-eng-fra), but it only produces noise (not intelligible speech sounds) after cleaning about 3k worst datapoints from each. I run the scorer already 5 times, the last 2 times I deleted the worst 1k datapoints from each. I will try recreating the caches and deleting again the worst datapoints, but if you seek for another solution, please tell me @Flux9665 
![image](https://github.com/user-attachments/assets/db78315c-8840-475b-8890-d45faf28ebac)
![5000](https://github.com/user-attachments/assets/3e1664fd-8f17-472e-869c-1f062cbf5996)

**Comment by Flux9665** [comment]
Usually running the scorer just once is sufficient to get the model from a state where it fails to a state where it works, so that's strange. I see now that your runs only go to 5k steps, that means that the CFM decoder has not yet started training if you were training from scratch. But you're finetuning, right? 

I'm sorry but I don't see the issue right now. Can you maybe link me the the code of the pipeline you're using?

**Comment by AlexSteveChungAlvarez** [comment]
Yes, I'm finetuning.

```
"""
Example script for fine-tuning the pretrained model to your own data.

Comments in ALL CAPS are instructions
"""

import time
import torch
import wandb
from torch.utils.data import ConcatDataset

from Modules.ToucanTTS.ToucanTTS import ToucanTTS
from Modules.ToucanTTS.toucantts_train_loop_arbiter import train_loop
from Utility.corpus_preparation import prepare_tts_corpus
from Utility.path_to_transcript_dicts import *
from Utility.storage_config import MODELS_DIR
from Utility.storage_config import PREPROCESSING_DIR


def run(gpu_id, resume_checkpoint, finetune, model_dir, resume, use_wandb, wandb_resume_id, gpu_count):
    if gpu_id == "cpu":
        device = torch.device("cpu")
    else:
        device = torch.device("cuda")
    assert gpu_count == 1  # distributed finetuning is not supported

    # IF YOU'RE ADDING A NEW LANGUAGE, YOU MIGHT NEED TO ADD HANDLING FOR IT IN Preprocessing/TextFrontend.py

    print("Preparing")

    if model_dir is not None:
        save_dir = model_dir
    else:
        save_dir = os.path.join(MODELS_DIR, "ToucanTTS_Opensource_HQ_3")  # RENAME TO SOMETHING MEANINGFUL FOR YOUR DATA
    os.makedirs(save_dir, exist_ok=True)

    all_train_sets = list()  # YOU CAN HAVE MULTIPLE LANGUAGES, OR JUST ONE. JUST MAKE ONE ConcatDataset PER LANGUAGE AND ADD IT TO THE LIST.
    train_samplers = list()

    # ========================
    # =    English Data      =
    # ========================
    english_datasets = list()
    english_datasets.append(prepare_tts_corpus(transcript_dict=build_path_to_transcript_cv_english(),#chunks[index],
                                                      corpus_dir=os.path.join(PREPROCESSING_DIR, f"cv_english"),
                                                      lang="eng"))

    all_train_sets.append(ConcatDataset(english_datasets))
    
    # ========================
    # =    French Data       =
    # ========================
    french_datasets = list()
    french_datasets.append(prepare_tts_corpus(transcript_dict=build_path_to_transcript_cv_french(),#chunks[index],
                                                      corpus_dir=os.path.join(PREPROCESSING_DIR, f"cv_french"),
                                                      lang="fra"))
    
    all_train_sets.append(ConcatDataset(french_datasets))    
    
    # ========================
    # =    Spanish Data      =
    # ========================
    spanish_datasets = list()
    spanish_datasets.append(prepare_tts_corpus(transcript_dict=build_path_to_transcript_cv_spanish(),#chunks[index],
                                                      corpus_dir=os.path.join(PREPROCESSING_DIR, f"cv_spanish"),
                                                      lang="spa"))
    
    all_train_sets.append(ConcatDataset(spanish_datasets))

    model = ToucanTTS()

    for train_set in all_train_sets:
        train_samplers.append(torch.utils.data.RandomSampler(train_set))

    if use_wandb:
        wandb.init(
            name=f"{__name__.split('.')[-1]}_{time.strftime('%Y%m%d-%H%M%S')}" if wandb_resume_id is None else None,
            id=wandb_resume_id,  # this is None if not specified in the command line arguments.
            resume="must" if wandb_resume_id is not None else None)

    print("Training model")
    train_loop(net=model,
               datasets=all_train_sets,
               device=device,
               save_directory=save_dir,
               batch_size=12,  # YOU MIGHT GET OUT OF MEMORY ISSUES ON SMALL GPUs, IF SO, DECREASE THIS.
               eval_lang="fra",  # THE LANGUAGE YOUR PROGRESS PLOTS WILL BE MADE IN
               warmup_steps=500,
               lr=1e-5,  # if you have enough data (over ~1000 datapoints) you can increase this up to 1e-4 and it will still be stable, but learn quicker.
               # DOWNLOAD THESE INITIALIZATION MODELS FROM THE RELEASE PAGE OF THE GITHUB OR RUN THE DOWNLOADER SCRIPT TO GET THEM AUTOMATICALLY
               path_to_checkpoint=os.path.join(MODELS_DIR, "ToucanTTS_Meta", "best.pt") if resume_checkpoint is None else resume_checkpoint,
               fine_tune=True if resume_checkpoint is None and not resume else finetune,
               resume=resume,
               steps=5000,
               use_wandb=use_wandb,
               train_samplers=train_samplers,
               gpu_count=1)
    if use_wandb:
        wandb.finish()
```

These are the alignment plots:
french:
![image](https://github.com/user-attachments/assets/e4178b7e-b3e2-44d0-a551-a107d87b25ea)

english:
![image](https://github.com/user-attachments/assets/2b300fdf-b495-44e9-bf06-dcd1159b8782)

spanish:
![image](https://github.com/user-attachments/assets/e078750e-7eb6-4a9d-a6ef-733a22240830)

**Comment by Flux9665** [comment]
All looks good, I don't see a problem with the code. 

The MAS plot for the alignment of Spanish is however very off. Looking back at the loss curves you posted earlier, I think that there is still some problem with the data. Maybe the alignments are not very good because the aligner model was trained on the data before the cleaning was applied and there were still very broken samples in the data. The stochastic loss is increasing a lot, maybe we should just not tune the CFM at all in case the data has problems.

You could try to see if the model still works if the CFM is not used. For that, you can replace in this line in the InferenceToucanTTS

https://github.com/DigitalPhonetics/IMS-Toucan/blob/9f45a30fd15b6f1551b6e90c1af60fff8890c9ce/Modules/ToucanTTS/InferenceToucanTTS.py#L274

the `refined_codec_frames` with `preliminary_spectrogram` and then see how it sounds.

**Comment by AlexSteveChungAlvarez** [comment]
> the `refined_codec_frames` with `preliminary_spectrogram` and then see how it sounds.

It didn't work neither.

>I think that there is still some problem with the data.

I'm running another tool to clean the data more, and when running it, I found out that I was using the deprecated limit_to_n function of the previous version during the cache building. It always limits the whole data to 40k datapoints, do you think this could be the problem? 

The tool I'm running to clean the data predicts the MOS of the file and if it has less value from the average MOS it's taken away. Furthermore, if the MOS is "nan", then it deletes the audio from the original dataset. So I have two options now, or I run the entire training since building the caches without the limit_to_n function or I wait until this tool cleans the whole data. @Flux9665 Do you think running without the deprecated function would solve the issue?

**Comment by Flux9665** [comment]
No, this would not fix the problem. I have no idea what the problem actually is, but it's not the amount of datapoints. 

Also, the quality of the samples is likely not the issue, so I don't think that the MOS prediction will fix it. Considering the MAS plot for Spanish, the issue is likely that the text and the speech don't always match. So there are datapoints where the transcription contains things that are not being said, or there are things being said that are not part of the transcription.

**Comment by AlexSteveChungAlvarez** [comment]
> No, this would not fix the problem. I have no idea what the problem actually is, but it's not the amount of datapoints.
> 
> Also, the quality of the samples is likely not the issue, so I don't think that the MOS prediction will fix it. Considering the MAS plot for Spanish, the issue is likely that the text and the speech don't always match. So there are datapoints where the transcription contains things that are not being said, or there are things being said that are not part of the transcription.

OK, the tool I'm running also does ASR and returns new transcriptions with the audio, so I hope this new dataset doesn't have problems. I'll get first a subset of the hole data to test if this works.

**Comment by AlexSteveChungAlvarez** [comment]
Finally I can hear voice outputs, though it seems the quality of the data wasn't so good because the output of the finetuned model is not better than the base one.
